IBX-5706: DDEV usage (#1983)
ibexa/documentation-developer: docs/getting_started/install_using_ddev.md

> # Install using DDEV

Author: Added in 0d62b872365b1ad5e6574bb51cdc79a22061d93b

@DominikaK please review

This guide provides a step-by-step walkthrough of installing [[= product_name =]] using [DDEV](https://ddev.com/).
DDEV is an open-source tool that simplifies the process of setting up local PHP development environments.

## System requirements

For a successful installation, ensure you have the following software installed:

- [Docker](https://docs.docker.com/get-docker/)
- [DDEV](https://ddev.readthedocs.io/en/latest/users/install/ddev-installation/)

## Installation steps

### 1. Create a DDEV project directory

Start by creating a directory for your DDEV project. Navigate to your desired location in the terminal, then use the following command to create a new directory:

```bash
mkdir my-ddev-project && cd my-ddev-project
```

Replace `my-ddev-project` with your desired directory name.

### 2. Configure DDEV

#### Configure PHP version and document root

Next, configure your DDEV environment using the command below:

```bash
ddev config --project-type=php --php-version 8.1 --docroot=public --create-docroot
```

This command sets the project type to PHP, the PHP version to 8.1, the document root to `public` directory, and creates the document root.

> #### Optional: Switch to Apache

Author: Done in 54235a3c9e9918659a4c6509d63b0f1d9e635e0a

#### Configure database connection

Now, configure the database connection for your Ibexa DXP project. Depending on your database of choice (MySQL or PostgreSQL), use the appropriate command below.

!!! note

    Those commands will set a `DATABASE_URL` environment variable inside the container which override [the .env's one](install_ibexa_dxp.md#change-installation-parameters).

#### MariaDB

```bash
ddev config --web-environment-add DATABASE_URL=mysql://db:db@db:3306/db
```

#### PostgreSQL

```bash
ddev config --web-environment-add DATABASE_URL=postgresql://db:db@db:5432/db
```

#### Optional: Enable Mutagen

If you're using macOS or Windows, you might wish to enable [Mutagen](https://ddev.readthedocs.io/en/latest/users/install/performance/#mutagen) for performance. You can do this by running the following command:

```bash
ddev config --mutagen-enabled
```

See [DDEV performance documentation](https://ddev.readthedocs.io/en/latest/users/install/performance/) for more.

#### Optional: Change port mapping

By default, DDEV will use ports 80 and 443.

> You can [set different ports](https://ddev.readthedocs.io/en/latest/users/usage/troubleshooting/#method-2-fix-port-conflicts-by-configuring-your-project-to-use-different-ports) with a command like the following:

Author: They are ports from host to container, internally mapped to 80 and 443.
Those ports can be already used. By another DDEV project for example. On my Mac, port 443 can't be used by default `Ports are not available: exposing port TCP 127.0.0.1:443 -> 0.0.0.0:0`. It's easier to just use another port than trying to free this port (that could be refused for security reason, I don't know).

> You can [set different ports](https://ddev.readthedocs.io/en/latest/users/usage/troubleshooting/#method-2-fix-port-conflicts-by-configuring-your-project-to-use-different-ports) with a command like the following:

Reviewer: I think another DDEV project is a bad example since DDEV is using the router.
At least as far as I'm aware.

> You can [set different ports](https://ddev.readthedocs.io/en/latest/users/usage/troubleshooting/#method-2-fix-port-conflicts-by-configuring-your-project-to-use-different-ports) with a command like the following:

Author: @ibexa-yuna You are right, DDEV has no port conflict between DDEV projects thx to its ddev-router.

Still, the conflict can be with an other web server already running, like, for example, the one used before migrating to DDEV.

```bash
ddev config --http-port=8080 --https-port=8443
```

### 3. Start DDEV

Proceed by starting your DDEV environment with the following command:

```bash
ddev start
```

!!! tip

>     If you forgot a configuration, you can still use `ddev config` but you got to restart afterward using `ddev restart`.

Author: @DominikaK Yes. I just remove a outnumbered "to"
```suggestion
    If you forgot some part of configuration, you can set it by using `ddev config` later, but afterwards you have to restart DDEV using `ddev restart`.
```

### 4. Composer authentication

If you're installing the Commerce, Experience, or Content edition of Ibexa DXP, you'll need to [set up authentication tokens](install_ibexa_dxp.md#set-up-authentication-tokens) by modifying the Composer configuration.
It must be done **after** executing the `ddev start` command as it will be run inside the container.

```bash
ddev composer config --global http-basic.updates.ibexa.co <installation-key> <token-password>
```

Replace `<installation-key>` and `<token-password>` with your actual installation key and token password, respectively.

If you want to reuse an existing `auth.json`file, see [Alternatives / Using an auth.json](#using-an-authjson).

### 5. Create Ibexa DXP project

Once DDEV is running, use Composer to create a new Ibexa DXP project. Remember to replace `<edition>` and `<version>` with your desired edition and versions respectively.

```bash

> ddev composer create ibexa/<edition>-skeleton:<version>

Author: It's a trap.

```
% ddev help composer;
[…]
Available Commands:
  create         Executes 'composer create-project' within the web container with the arguments and flags provided
  create-project Unsupported, use `ddev composer create` instead
```

```

### 6. Install the DXP and its database

Once you've made this change, you can proceed to install Ibexa DXP.

```bash
ddev php bin/console ibexa:install ibexa-<edition>
ddev php bin/console ibexa:graphql:generate-schema
```

### 7. Open browser

Once the above steps are completed, open your Ibexa DXP webpage by running the `ddev launch` command.

!!! Note

    The project's URL was also given in `ddev start` output.

### 8. Start using Ibexa DXP

You can now start using Ibexa DXP and implement your own website on the platform.

> You can follow the [first usual steps](first_steps.md), the [Beginner tutorial](../tutorials/beginner_tutorial/beginner_tutorial.md) or partner agreements trainings.

Author: ```suggestion
```

The configuration and code in the DDEV project directory can be edited.
You can use the command you'll cross in the documentation by prefixing them with `ddev exec` or by opening a terminal inside the container using `ddev ssh`.
For example, if a guideline invites you to run `php bin/console cache:clear`, you can do it in the DDEV container by following one of those two ways:

> - running `ddev php bin/console cache:clear`

Author: I would say no. An alias lover like you and me will take the initiative to create his/her own way. To explain `alias`, bashrc, zshrc, other shell, or other "alway available alias" strategies seems out of the scope.

- entering `ddev ssh` and run after the new prompt `php bin/console cache:clear`

## Going further

> ### Alternatives

Author: @DominikaK 
To expose a bit more what the alternative are, "Alternatives" can be removed or merge, and it's sub-section level-down and show in the right TOC:

| Actual | Proposal
|-------|----------
|## Going further<br>### Alternatives | ## Going further and alternatives
|#### Using `auth.json`|### Using `auth.json`
|#### Using Dotenv|### Using Dotenv
|### Webserver configuration|
|#### Nginx Server Blocks|### Nginx Server Blocks
|#### Apache Virtual Host|### Apache Virtual Host
|##### Manual|#### Manual
|##### Scripted|##### Scripted
|### Run an already existing project|### Run an already existing project
|### Mimicking a production environment|### Mimicking a production environment
|#### Hostnames and domains|#### Hostnames and domains
|#### Cluster or Ibexa Cloud|#### Cluster or Ibexa Cloud

> ### Alternatives

Author: > How about something like "Other options for configuration"? Ideally the heading should _what_ these are alternatives are for

@DominikaK Thx for this re-focus. I like your title.

To remove `### Webserver configuration` doesn't work well. To use tabs for Nginx/Apache doesn't work either. So, I propose to re-evaluate level for the first sub-sections.

| Actual | Proposal
|-------|----------
|## Going further<br>### Alternatives | ## Other options for configuration
|#### Using `auth.json`|### Using `auth.json`
|#### Using Dotenv|### Using Dotenv
|### Webserver configuration|### Webserver configuration
|#### Nginx Server Blocks|#### Nginx Server Blocks
|#### Apache Virtual Host|#### Apache Virtual Host
|##### Manual|##### Manual
|##### Scripted|##### Scripted
|### Run an already existing project|### Run an already existing project
|### Mimicking a production environment|### Mimicking a production environment
|#### Hostnames and domains|#### Hostnames and domains
|#### Cluster or Ibexa Cloud|#### Cluster or Ibexa Cloud

> ### Alternatives

Author: Last proposal done in 50d028f35ae53348a96591cc6aaa8f5d404069e2

| `%TIMEOUT%`         | `0`                          |
| `%FASTCGI_PASS%`    | `unix:/var/run/php-fpm.sock` |

For example, it can be done with `sed`:

```bash
sed -i 's/%IP_ADDRESS%/*/' .ddev/apache/apache-site.conf
sed -i 's/%PORT%/80/' .ddev/apache/apache-site.conf
sed -i 's/%BASEDIR%/\/var\/www\/html/' .ddev/apache/apache-site.conf
sed -i 's/%BODY_SIZE_LIMIT%/0/' .ddev/apache/apache-site.conf
sed -i 's/%TIMEOUT%/0/' .ddev/apache/apache-site.conf
sed -i 's/%FASTCGI_PASS%/unix:\/var\/run\/php-fpm.sock/' .ddev/apache/apache-site.conf
```

Finally, restart the project:

```bash
ddev restart
```

##### Scripted

Generate the Virtual Host using https://github.com/ibexa/docker/blob/main/scripts/vhost.sh:

```bash
curl -O https://raw.githubusercontent.com/ibexa/docker/main/scripts/vhost.sh
bash vhost.sh --template-file=vendor/ibexa/post-install/resources/templates/apache2/vhost.template \
  --ip='*' \
  --host-name='my-ddev-project.ddev.site' \
  --host-alias='*.ddev.site' \
  --basedir='/var/www/html' \
  --sf-env=dev \
  > .ddev/apache/apache-site.conf
rm vhost.sh
```

Then, restart the project:

```bash
ddev restart
```

### Run an already existing project

> - Instead of creating an empty directory like in _[1. Create a DDEV project directory](#1-create-a-ddev-project-directory)_, a directory already containing an Ibexa DXP project is used.

Author: Done in fe3d416499fa3f1e9d6cf8975a4663fd2e89fc36

> - In _[2. Configure DDEV / Configure PHP version and document root](#configure-php-version-and-document-root)_, no need to create the Document root, `--create-docroot` mustn't be used.

Author: Done in fe3d416499fa3f1e9d6cf8975a4663fd2e89fc36

> - Instead of `ddev composer create` from _[5. Create Ibexa DXP project](#5-create-ibexa-dxp-project)_, use solely `ddev composer install`.

Reviewer 2: ```suggestion
- Instead of `ddev composer create-project` from _[5. Create Ibexa DXP project](#5-create-ibexa-dxp-project)_, use solely `ddev composer install`.
```

Author: idem, `ddev composer create-project` is not supported. Maybe `ddev exec composer create-project` but I never tested it preferring the buit-in short-cut.

> - Instead of `ddev composer create` from _[5. Create Ibexa DXP project](#5-create-ibexa-dxp-project)_, use solely `ddev composer install`.

Author: Done in fe3d416499fa3f1e9d6cf8975a4663fd2e89fc36

- [Ibexa data migration](../content_management/data_migration/importing_data.md) or [`ddev import-db`](https://ddev.readthedocs.io/en/latest/users/usage/commands/#import-db) can be used to populate the database.
- Content's binary files can be copied into `public/var`.

> The following example run on DDEV a [version controlled project](https://doc.ibexa.co/en/latest/getting_started/install_ibexa_dxp/#add-project-to-version-control), a local clone of a remote Git repository:

Author: Hyphen added in d6b38f5175054dff69c9ab3f5f2f3d5f667b7fd5

The rest is outdated.

```bash
git clone <repository> my-ddev-project && cd my-ddev-project
.ddev/ >> .gitignore
ddev config --project-type=php --php-version 8.1 \
  --docroot=public \
  --web-environment-add DATABASE_URL=mysql://db:db@db:3306/db \
  --http-port=8080 --https-port=8443
ddev start
ddev composer config --global http-basic.updates.ibexa.co <installation-key> <token-password>
ddev php bin/console ibexa:migrations:migrate --file=project_content_types.yaml
ddev composer install
ddev launch
```

> Notice that the example choose to `.gitignore` the whole `.ddev/` directory. It can go otherwise. Some DDEV configs can be shared among developers. For example, a common `.ddev/config.yaml` can be committed for everyone and [locally extended or override](https://ddev.readthedocs.io/en/latest/users/extend/customization-extendibility/#extending-configyaml-with-custom-configyaml-files). 

Author: Done in 7b9728810712783a4b0216f0199de4808f9c93da

I changed "you can also version it" with "you can also version it" because there might be some unwanted things in it, like the `.ddev/homeadditions/.composer/auth.json`.

```bash
ddev composer create ibexa/<edition>-skeleton:<version>
```

### 6. Install the DXP and its database

Once you've made this change, you can proceed to install Ibexa DXP.

```bash
ddev php bin/console ibexa:install ibexa-<edition>
ddev php bin/console ibexa:graphql:generate-schema
```

### 7. Open browser

Once the above steps are completed, open your Ibexa DXP webpage by running the `ddev launch` command.

!!! tip

    You can also see the project URL in the `ddev start` output.

### 8. Start using Ibexa DXP

You can now start using Ibexa DXP and implement your own website on the platform.

You can edit the configuration and code in the DDEV project directory.
You can use commands listed in the documentation by prefixing them with `ddev exec` or by opening a terminal inside the container using `ddev ssh`.
For example, if a guideline invites you to run `php bin/console cache:clear`, you can do it in the DDEV container in one of the following ways:

- run `ddev php bin/console cache:clear`
- enter `ddev ssh` and run `php bin/console cache:clear` after the new prompt

## Other options for configuration

> DDEV offers several ways to achieve a same thing, offering different levels of flexibility or adaptability to your development environment.

Reviewer 3: ```suggestion
DDEV offers several ways to achieve the same thing, offering different levels of flexibility or adaptability to your development environment.
```

!!! tip

    Learn more about the [DDEV commands](https://ddev.readthedocs.io/en/latest/users/usage/commands/):
    

>     - by running [`ddev list`](https://ddev.readthedocs.io/en/latest/users/usage/commands/#list) to list them all,

Reviewer 3: @adriendupuis It lists all environments, so maybe it's good to precise in the description that it refers to environments not to listing all commands.

Author: @juskora I really meant to list the commands but got confused.

Can you check the following fix proposal, please?

```suggestion
    - by running [`ddev --help`](https://ddev.readthedocs.io/en/latest/users/usage/cli/#using-the-ddev-command) to list them all,
```

Reviewer 3: Yes, it's okay.

`cp <path-to-an>/auth.json .ddev/homeadditions/.composer`

Alternatively, the Composer global `auth.json` can be the DDEV global `auth.json` with the help of symbolic link:
`mkdir -p ~/.ddev/homeadditions/.composer && ln -s ~/.composer/auth.json ~/.ddev/homeadditions/.composer/auth.json`

If the DDEV project has already been started, you need to run `ddev restart`.

To use an `auth.json` file replaces the step [4. Composer authentication](#4-composer-authentication).

### Using Dotenv

Instead of using environment variables inside the container, a [`.env.local`](https://symfony.com/doc/5.4/configuration.html#overriding-environment-values-via-env-local) file can be added to the project.

The following shows `.env.local` configuration on the example of the database:

- Skip step [2. Configure DDEV / Configure database connection](#configure-database-connection).
- Modify step [5. Create Ibexa DXP project](#5-create-ibexa-dxp-project) to insert the database setting (Commerce and MariaDB are used in this example):
  ```bash
  ddev composer create ibexa/commerce-skeleton --no-install;
  echo "DATABASE_URL=mysql://db:db@db:3306/db" >> .env.local;
  ddev composer install;

>   ```

Reviewer 3: @adriendupuis  I think it's worth mentioning about this earlier, around this 2nd step which has to be skipped in the beginning of the procedure.

Author: Good remark. Mentioned in "Configure database connection" (addition in 238969ccb6ea437e9391493ff049c93aaf2ee665)

| `%PORT%`                | `80`                           |
| `%HOST_LIST%`           | `*.ddev.site`                  |
| `%BASEDIR%`             | `/var/www/html`                |
| `%BODY_SIZE_LIMIT_M%`   | `0`                            |
| `%TIMEOUT_S%`           | `0`                            |
| `%FASTCGI_PASS%`        | `unix:/var/run/php-fpm.sock`   |
| `%BINARY_DATA_HANDLER%` | empty string |

Because of path resolution inside DDEV's Nginx, you must replace one more thing: `ibexa_params.d` with `sites-enabled/ibexa_params.d`.

You can, for example, do it with `sed`:

```bash
sed -i 's/%PORT%/80/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%HOST_LIST%/*.ddev.site/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%BASEDIR%/\/var\/www\/html/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%BODY_SIZE_LIMIT_M%/0/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%TIMEOUT_S%/60s/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%FASTCGI_PASS%/unix:\/var\/run\/php-fpm.sock/' .ddev/nginx_full/ibexa.conf;
sed -i 's/%BINARY_DATA_HANDLER%//' .ddev/nginx_full/ibexa.conf;
sed -i 's/ibexa_params.d/sites-enabled\/ibexa_params.d/' .ddev/nginx_full/ibexa.conf;
```

#### Apache Virtual Host

> To set the Apache Virtual Host, `.ddev/apache/apache-site.conf` will be override with Ibexa DXP's config. You can do it manually or using a script.

Reviewer 3: ```suggestion
To set the Apache Virtual Host, `.ddev/apache/apache-site.conf` will be overriden with Ibexa DXP's config. You can do it manually or using a script.
```

Author: [What a conjugation…](https://en.wiktionary.org/wiki/override#Verb)

```suggestion
To set the Apache Virtual Host, `.ddev/apache/apache-site.conf` will be overridden with Ibexa DXP's config. You can do it manually or using a script.
```

- In [2. Configure DDEV / Configure PHP version and document root](#configure-php-version-and-document-root), don't create the Document root, remove the `--create-docroot` option.
- Instead of `ddev composer create`, in [5. Create Ibexa DXP project](#5-create-ibexa-dxp-project) use only `ddev composer install`.
- [Ibexa data migration](../content_management/data_migration/importing_data.md) or [`ddev import-db`](https://ddev.readthedocs.io/en/latest/users/usage/commands/#import-db) can be used to populate the database.

### Hostnames and domains

If the local project needs to answer to real production domains (for example, to use the existing [hostname to SiteAccess](../multisite/siteaccess/siteaccess_matching.md#maphost) or [hostname element to SiteAccess](../multisite/siteaccess/siteaccess_matching.md#hostelement) mappings), you can use [additional hostnames](https://ddev.readthedocs.io/en/latest/users/extend/additional-hostnames/).

!!! warning

    As this feature will modify domain resolution, the real website may be unreachable until the `hosts` file is manually cleaned.

> 

Reviewer 3: @DominikaK  'warning' has different styling from others: tip, caution, notice, is it by design?

Author: @juskora 

Thx. I meant "caution."

Author: ```suggestion
!!! caution

    As this feature will modify domain resolution, the real website may be unreachable until the `hosts` file is manually cleaned.

```

```

This command sets the project type to PHP, the PHP version to 8.1, the document root to `public` directory, and creates the document root.

#### Switch to Apache (optional)

By default, DDEV uses Nginx.

To use Apache instead, run the following command:

```bash
ddev config --webserver-type=apache-fpm
```

#### Use another database type (optional)

By default, DDEV uses MariaDB.

To use PostgreSQL instead, run the following command:

```bash
ddev config --database=postgres:15
```

To use MySQL instead, run the following command:

```bash
ddev config --database=mysql:8.0
```

Other version of MariaDB, Mysql or PostgreSQL can be used. See [DDEV database types documentation](https://ddev.readthedocs.io/en/latest/users/extend/database-types/) for available version ranges.

#### Configure database connection

Now, configure the database connection for your Ibexa DXP project. Depending on your database of choice (MySQL or PostgreSQL), use the appropriate command below.

!!! note

    Those commands will set a `DATABASE_URL` environment variable inside the container which overrides [the variable from `.env`](install_ibexa_dxp.md#change-installation-parameters).

>     If you wan't to use `.env.local` file instead of server-level environment variables, see [Using dotenv](#using-dotenv).

Reviewer 3: Yes, looks great! :)

Next, you need to [set up authentication tokens](install_ibexa_dxp.md#set-up-authentication-tokens) by modifying the Composer configuration.
You must run the following command **after** executing `ddev start`, because the command runs inside the container.
Replace `<installation-key>` and `<token-password>` with your actual installation key and token password, respectively.

```bash
ddev composer config --global http-basic.updates.ibexa.co <installation-key> <token-password>
```

This authentication won't persist if the project is restarted (by `ddev restart` or `ddev composer create`). You can back up the authentication file (`auth.json`) by using the following command:

```bash
ddev exec "mkdir -p .ddev/homeadditions/.composer; cp ~/.composer/auth.json .ddev/homeadditions/.composer"
```

If you want to reuse an existing `auth.json` file, see [Using an auth.json](#using-an-authjson).

### 5. Create Ibexa DXP project

Once DDEV is running, use Composer to create a new Ibexa DXP project. Remember to replace `<edition>` and `<version>` with your desired edition and version respectively.

```bash
ddev composer create ibexa/<edition>-skeleton:<version>
```

!!! tip

>     You may add the Symfony Debug bundle for its var dumper enhancement which allows to avoid getting out of memory error when dumping objects containing circular references.

Author: Trying to merge @dabrt and @adamwojs suggestions, I come to this:
```suggestion
    Consider adding the Symfony Debug bundle to benefit from its VarDumper enhancement that avoids getting the "out of memory" error when dumping objects containing circular references.
```

Author: ```suggestion
    Consider adding the Symfony Debug bundle which fixes memory outage when dumping objects with circular references.
```

---
description: Install Ibexa DXP with Docker and DDEV to use it for development.
---

# Install using DDEV

This guide provides a step-by-step walkthrough of installing [[= product_name =]] using [DDEV](https://ddev.com/).
DDEV is an open-source tool that simplifies the process of setting up local PHP development environments.

## Requirements

Before you start the installation, ensure you have the following software installed:

- [Docker](https://docs.docker.com/get-docker/)
- [DDEV](https://ddev.readthedocs.io/en/latest/users/install/ddev-installation/)

## Installation

### 1. Create a DDEV project directory

Start by creating a directory for your DDEV project by using the following command to create a new directory:

```bash
mkdir my-ddev-project && cd my-ddev-project
```

> Replace `my-ddev-project` with your desired directory name.

Author: Done in a75667c07f3b82d736f15ed0ee199a1a7dee77c3

Couldn't be applied directly from Github because of triple backquotes conflict between code and suggestion.

#### Configure PHP version and document root

Next, configure your DDEV environment using the command below:

```bash
ddev config --project-type=php --php-version 8.1 --docroot=public --create-docroot
```

This command sets the project type to PHP, the PHP version to 8.1, the document root to `public` directory, and creates the document root.

#### Switch to Apache (optional)

By default, DDEV uses Nginx.

To use Apache instead, run the following command:

```bash
ddev config --webserver-type=apache-fpm
```

#### Use another database type (optional)

By default, DDEV uses MariaDB.

To use PostgreSQL instead, run the following command:

```bash
ddev config --database=postgres:15
```

To use MySQL instead, run the following command:

```bash
ddev config --database=mysql:8.0
```

Other version of MariaDB, Mysql or PostgreSQL can be used. See [DDEV database types documentation](https://ddev.readthedocs.io/en/latest/users/extend/database-types/) for available version ranges.

#### Configure database connection

Now, configure the database connection for your Ibexa DXP project. Depending on your database of choice (MySQL or PostgreSQL), use the appropriate command below.

!!! note

>     Those commands will set a `DATABASE_URL` environment variable inside the container which overrides [the variable from `.env`](install_ibexa_dxp.md#change-installation-parameters).

Reviewer 4: ```suggestion
    Those commands set a `DATABASE_URL` environment variable inside the container which overrides [the variable from `.env`](install_ibexa_dxp.md#change-installation-parameters).
```

Author: There is a bug with Vale here. If I remove the indentation, I obtain the expected `Ibexa.FutureTense` warning "Use present tense instead of future."

#### Change port mapping (optional)

By default, DDEV uses ports 80 and 443.
You can [set different ports](https://ddev.readthedocs.io/en/latest/users/usage/troubleshooting/#method-2-fix-port-conflicts-by-configuring-your-project-to-use-different-ports) with a command like the following:

```bash
ddev config --http-port=8080 --https-port=8443
```

### 3. Start DDEV

Proceed by starting your DDEV environment with the following command:

```bash
ddev start
```

!!! tip

    If you forgot some part of configuration, you can set it by using `ddev config` later, but afterwards you have to restart DDEV using `ddev restart`.

### 4. Composer authentication

Next, you need to [set up authentication tokens](install_ibexa_dxp.md#set-up-authentication-tokens) by modifying the Composer configuration.
You must run the following command **after** executing  `ddev start`, because the command run inside the container.

```bash
ddev composer config --global http-basic.updates.ibexa.co <installation-key> <token-password>
```

Replace `<installation-key>` and `<token-password>` with your actual installation key and token password, respectively.

> 

Author: Done in a75667c07f3b82d736f15ed0ee199a1a7dee77c3

This authentication won't persist if the project is restarted (by `ddev restart` or `ddev composer create`). You can back up the authentication file (`auth.json`) using the following command:

```bash
ddev exec "mkdir -p .ddev/homeadditions/.composer; cp ~/.composer/auth.json .ddev/homeadditions/.composer"

> ```

Author: The idea is to not use Composer on the host to generate an `auth.json` file. In those steps, only Docker and DDEV are required, not PHP, not Composer. The user is not supposed to already have an `auth.json` and even less to use Composer to generate one on the host. This hack is the best way to generate an `auth.json` inside the container and back it up on the host.

If you want to reuse an existing `auth.json` file, see [Using an auth.json](#using-an-authjson).

### 5. Create Ibexa DXP project

Once DDEV is running, use Composer to create a new Ibexa DXP project. Remember to replace `<edition>` and `<version>` with your desired edition and versions respectively.

```bash
ddev composer create ibexa/<edition>-skeleton:<version>
```

!!! tip

>     You may add the Symfony Debug bundle for its var dumper enhancement which can avoid getting out of memory when dumping huge or circular variables.

Author: Suggestion updated in a new suggestion.

>     `ddev composer require --dev symfony/debug-bundle`

Author: @adamwojs Is this introduction of the `symfony/debug-bundle` accurate?

Reviewer 5: This is more precise statement: 

"You may add the Symfony Debug bundle for its var dumper enhancement which allows to  avoid getting out of memory error when dumping objects containing circular references."

Author: ```suggestion
    You may add the Symfony Debug bundle for its var dumper enhancement which allows to avoid getting out of memory error when dumping objects containing circular references.
    `ddev composer require --dev symfony/debug-bundle`
```

- Instead of `ddev composer create`, in [5. Create Ibexa DXP project](#5-create-ibexa-dxp-project) use only `ddev composer install`.
- [Ibexa data migration](../content_management/data_migration/importing_data.md) or [`ddev import-db`](https://ddev.readthedocs.io/en/latest/users/usage/commands/#import-db) can be used to populate the database.

### Hostnames and domains

If the local project needs to answer to real production domains (for example, to use the existing [hostname to SiteAccess](../multisite/siteaccess/siteaccess_matching.md#maphost) or [hostname element to SiteAccess](../multisite/siteaccess/siteaccess_matching.md#hostelement) mappings), you can use [additional hostnames](https://ddev.readthedocs.io/en/latest/users/extend/additional-hostnames/).

!!! caution

    As this feature will modify domain resolution, the real website may be unreachable until the `hosts` file is manually cleaned.

### Cluster or Ibexa Cloud

DDEV can be useful to locally simulate a production cluster.

- See [clustering using DDEV](clustering_using_ddev.md) to add Elasticsearch, Solr, Redis or Memcached to your DDEV installation.

> - See [Ibexa Cloud and DDEV](ibexa_cloud_and_ddev.md) to locally run an Ibexa DXP project using DDEV.

Reviewer 4: ```suggestion
- See [Ibexa Cloud and DDEV](ibexa_cloud_and_ddev.md) to locally run an Ibexa DXP project by using DDEV.
```

| `%FASTCGI_PASS%`    | `unix:/var/run/php-fpm.sock` |

You can, for example, do it with `sed`:

```bash
sed -i 's/%IP_ADDRESS%/*/' .ddev/apache/apache-site.conf
sed -i 's/%PORT%/80/' .ddev/apache/apache-site.conf
sed -i 's/%BASEDIR%/\/var\/www\/html/' .ddev/apache/apache-site.conf
sed -i 's/%BODY_SIZE_LIMIT%/0/' .ddev/apache/apache-site.conf
sed -i 's/%TIMEOUT%/0/' .ddev/apache/apache-site.conf
sed -i 's/%FASTCGI_PASS%/unix:\/var\/run\/php-fpm.sock/' .ddev/apache/apache-site.conf
```

Finally, restart the project:

```bash
ddev restart
```

##### Scripted

Generate the Virtual Host using https://github.com/ibexa/docker/blob/main/scripts/vhost.sh:

```bash
curl -O https://raw.githubusercontent.com/ibexa/docker/main/scripts/vhost.sh
bash vhost.sh --template-file=vendor/ibexa/post-install/resources/templates/apache2/vhost.template \
  --ip='*' \
  --host-name='my-ddev-project.ddev.site' \
  --host-alias='*.ddev.site' \
  --basedir='/var/www/html' \
  --sf-env=dev \
  > .ddev/apache/apache-site.conf
rm vhost.sh
```

Then, restart the project:

```bash
ddev restart
```

### Run an already existing project

To run an existing project, you'll need to

> - configure the DDEV project

Author: It goes well with the numbered steps of the clean install.

Done in 2eab6f24c529bbf5e483c7bdaa23549d8535b1d1

- start the DDEV project

> - add Composer authentication if it's a Content, Experience or Commerce project

Author: ```suggestion
- add Composer authentication
```

- install dependencies packages using Composer
- populate the contents, which could be
  - getting a clean database using ddev `php bin/console ibexa:install ibexa-<edition>` then adding some data using [Ibexa data migration](../content_management/data_migration/importing_data.md)
  - injecting a dump using [`ddev import-db`](https://ddev.readthedocs.io/en/latest/users/usage/commands/#import-db) and copying related binary files into `public/var`

This example will run an existing project and have the right content structure but no content. This following script sequence will

> - clone [version-controlled project](install_ibexa_dxp.md#add-project-to-version-control) from a Git repository,

Reviewer 6: Is it possible to leave comments in bash? Then maybe moving each bullet point above the relevant line would make the example easier to understand and de-bullet point the text itself.

Author: Yes. Can be less readable but here the previous ordered list already introduce the strategy. Done in 2eab6f24c529bbf5e483c7bdaa23549d8535b1d1

- open the project in the default brother which should display the default SiteAccess frontpage.

```bash
git clone <repository> my-ddev-project && cd my-ddev-project
.ddev/ >> .gitignore
ddev config --project-type=php --php-version 8.1 \
  --docroot=public \
  --web-environment-add DATABASE_URL=mysql://db:db@db:3306/db \
  --http-port=8080 --https-port=8443
ddev start
ddev composer config --global http-basic.updates.ibexa.co <installation-key> <token-password>
ddev composer install
ddev php bin/console ibexa:install ibexa-<edition>
ddev php bin/console ibexa:migrations:migrate --file=project_content_types.yaml
ddev php bin/console ibexa:graphql:generate-schema
ddev launch
```

Notice that the example adds the whole `.ddev/` directory to `.gitignore`, but you can also version parts of it.
Some DDEV configs can be shared among developers. For example, a common `.ddev/config.yaml` can be committed for everyone and [locally extended or overridden](https://ddev.readthedocs.io/en/latest/users/extend/customization-extendibility/#extending-configyaml-with-custom-configyaml-files).

> Compared to running a clean install like described in _[Installation steps](#installation-steps)_:

Reviewer 6: ```suggestion
Compared to running a clean install like described in [Installation steps](#installation-steps):
```
